Description: Ignore SIGPIPE in the root/listener server context
 Misbehaving client can bring the whole nbd-server down, because
 nbd-server blindly believes that nothing goes wrong during negotation
 procedure. This patch makes nbd-server to ignore SIGPIPE to avoid
 unexpected termination.

 See http://sourceforge.net/mailarchive/message.php?msg_id=30410146

Author: Tuomas Räsänen <tuomasjjrasanen@opinsys.fi>

--- nbd-3.2.orig/nbd-server.c
+++ nbd-3.2/nbd-server.c
@@ -2203,6 +2203,7 @@ handle_connection(GArray *servers, int n
 		sigemptyset(&newset);
 		sigaddset(&newset, SIGCHLD);
 		sigaddset(&newset, SIGTERM);
+		sigaddset(&newset, SIGPIPE);
 		sigprocmask(SIG_BLOCK, &newset, &oldset);
 		if ((pid = fork()) < 0) {
 			msg3(LOG_INFO,"Could not fork (%s)",strerror(errno)) ;
@@ -2221,6 +2222,7 @@ handle_connection(GArray *servers, int n
 		/* child */
 		signal(SIGCHLD, SIG_DFL);
 		signal(SIGTERM, SIG_DFL);
+		signal(SIGPIPE, SIG_DFL);
 		sigprocmask(SIG_SETMASK, &oldset, NULL);
 
 		g_hash_table_destroy(children);
@@ -2466,6 +2468,7 @@ void setup_servers(GArray* servers) {
 	sa.sa_handler = sigchld_handler;
 	sigemptyset(&sa.sa_mask);
 	sigaddset(&sa.sa_mask, SIGTERM);
+	sigaddset(&sa.sa_mask, SIGPIPE);
 	sa.sa_flags = SA_RESTART;
 	if(sigaction(SIGCHLD, &sa, NULL) == -1)
 		err("sigaction: %m");
@@ -2473,9 +2476,18 @@ void setup_servers(GArray* servers) {
 	sa.sa_handler = sigterm_handler;
 	sigemptyset(&sa.sa_mask);
 	sigaddset(&sa.sa_mask, SIGCHLD);
+	sigaddset(&sa.sa_mask, SIGPIPE);
 	sa.sa_flags = SA_RESTART;
 	if(sigaction(SIGTERM, &sa, NULL) == -1)
 		err("sigaction: %m");
+
+	sa.sa_handler = SIG_IGN;
+	sigemptyset(&sa.sa_mask);
+	sigaddset(&sa.sa_mask, SIGCHLD);
+	sigaddset(&sa.sa_mask, SIGTERM);
+	sa.sa_flags = SA_RESTART;
+	if(sigaction(SIGPIPE, &sa, NULL) == -1)
+		err("sigaction: %m");
 }
 
 /**
